Climate React Auto Setup bug fix

seidnerj opened the pull request:
AC state setting change should not re-enable Climate React if Auto Setup is disabled.

seidnerj committed:
Climate React Auto Setup bug fix

AC state setting change should not re-enable Climate React if Auto Setup is disabled.

benwebbbenwebb commented on sensibo/StateHandler.js:
```diff
@@ -131,19 +129,27 @@ module.exports = (device, platform) => {
+						//       That means that even if a user turned Climate React off via the appropriate
+						//       switch, the next AC state setting that is changed will re-enable it.
+
+						device.state.smartMode.enabled = true
```
Hi @seidnerj 
Good explanation, however I feel that this still might be an issue...

If I've intentionally turned off smartMode (Climate React), but still have enableClimateReactAutoSetup=true, any change to my AC settings will re-enable smartMode. This doesn't seem intuitive to me.

I (personally) would expect that a setting/mode shouldn't enable/disable without my express action.

If I'm not wrong (and I easily could be, so let me know if so) changing something like swing while in fan mode could result in re-enabling smartMode, which could then turn on cooling or heating... right?

seidnerj replied:
Yes, you're correct. I agree it's not intuitive. 

The problem is it's not clear to me what should be the outcome of turning climate react off if it was turned on as a result of the auto setup feature. Should it essentially disable the feature? What behavior would one expect here?

Also - say auto setup is on and the user turned off climate react then turned it back on, what is the expected behavior?

There's also the issue of what happens when the user turns climate react on/off directly from Sensibo's app/website. 

The current flow provides consistent behavior regardless of the origin of such commands.

When a user turns on auto setup I believe it kind of implies to the user that the "system" will set climate react automatically, that means it will override any manual settings that the user sets. If the user doesn't want that then he/she can turn auto setup off. 

I guess we could also have a switch for auto setup, per AC device, which would only be created if auto setup is enabled in the plugin's configuration.

Anyways, I'm open to suggestions re what to do here - not sure what is best.

benwebbbenwebb replied:
Hi @seidnerj 

I've been thinking it over, my opinion is that when `enableClimateReactAutoSetup` is enabled, it should be limited in scope to adjusting only the ClimateReact settings (not enabling/disabling the ClimateReact mode). This would work similar to what you've implemented, if a user adjusts the temp / mode (heating/cooling) then that will be sent to the Sensibo API to adjust ClimateReact (SmartMode)*. If ClimateReact mode is currently enabled for the unit, send `device.state.smartMode.enabled = true`, if ClimateReact mode is currently disabled send `device.state.smartMode.enabled = false`.

A user can then enable/disable ClimateReact per unit via the Home app, or via the Sensibo app, and that will get synced (either via plugin PUT to Sensibo API or via GET Sensibo API request every 90 seconds) back to the other.

I think that will result in the most intuitive outcome for users.

\* Note: I'm thinking that adding an option to have the degree range on your ClimateReactAutoSetup feature configurable (1 as the default, but allow users to make it larger if they want) would be good, but I'll do that after this has been made live.

seidnerj replied:
I think that makes a lot of sense, I will make the appropriate adjustments. I agree re configuration of the degrees" delta", it is on my todo list.

Will update the PR when I'm done.

seidnerj commented:
Still haven't got around to doing this. Hopefully in the next few days.

seidnerj committed:
Climate React Auto Setup should not only set Climate React settings

Essentially, this commit simply removes code that forcefully enabled Climate React on every AC state change.

This feature will now be limited in scope to adjusting the Climate React _settings_ (but NOT enabling/disabling Climate React mode, this should be done by the user directly).

seidnerj commented:
I think that did it. It was easier than I thought once I took a better look at the code.

benwebbbenwebb commented:
Great, I’ll take a look over the weekend @seidnerj

benwebbbenwebb commented on sensibo/StateHandler.js:
```diff
@@ -131,22 +129,12 @@ module.exports = (device, platform) => {
 					// send state command to Sensibo
-
-					if (platform.enableClimateReactAutoSetup) {
-						await sensiboApi.setDeviceClimateReactState(device.id, sensiboNewClimateReactState)
```
Hi @seidnerj 

Don't you still need to send the new ClimateReactState to Sensibo when `enableClimateReactAutoSetup` = true?

Agree that `device.state.smartMode.enabled = true` needed to be removed (to prevent unexpectedly enabling ClimateReact) but - unless it's elsewhere - I think you still need to update the ClimateReact object whenever the AC state changes (to align with your AutoStep goals) and call `setDeviceClimateReactState(...)`

seidnerj replied:
If I understood your question correctly, the code that sets device.state.smartMode.enabled to true, is found in StateManager.js:

```
			// CLIMATE REACT
			ClimateReactEnabledSwitch: (state, callback) => {
				log.easyDebug(device.name, '(SET) - Climate React Enabled Switch:', state)
				const smartModeState = device.state.smartMode

				smartModeState.enabled = state

				// NOTE: we must set the 'smartMode' property directly (and NOT for example like so: device.state.smartMode.enabled = true),
				//       otherwise the StateHandler's setter code will not be executed and any changes will not take effect.
				device.state.smartMode = smartModeState

				callback()
			},
```

When the user toggles the switch device.state.smartMode is updated such that device.state.smartMode is true (or false, depending on the toggle). That, in return, triggers the following code in StateHandler.js:

```
			// Send Climate React state command and refresh state
			if (prop === 'smartMode') {
				try {
					log.easyDebug(`${device.name} - Setting Climate React state to ${value}`)
					const sensiboNewClimateReactState = unified.sensiboFormattedClimateReactState(device, state)

					sensiboApi.setDeviceClimateReactState(device.id, sensiboNewClimateReactState)
				} catch(err) {
					log('Error occurred! -> Climate React state did not change')
				}

				if (!platform.setProcessing) {
					platform.refreshState()
				} else {
					log.easyDebug('setProcessing is set to true, skipping state refresh due to Climate React set.')
				}

				return
			}
```

Which calls sensiboApi.setDeviceClimateReactState that posts new Climate React settings to Sensibo's API. The same happens when any relevant Climate React setting is set via calling the updateClimateReact function. 

For example, when setting CoolingThresholdTemperature (in StateManager.js) we then call updateClimateReact(device) which in turns triggers the aforementioned code in StateHandler.js the ends up triggering sensiboApi.setDeviceClimateReactState and posts the new Climate React settings to Sensibo's API.

Makes sense?

benwebbbenwebb replied:
Thanks @seidnerj (and sorry for the delay in replying)

I think I understand, but to be sure...

So, for example, because within `updateClimateReact(device)` (in StateManager.js) it finishes by doing this:
```
device.state.smartMode = smartModeState
```
that means the following is triggered (in StateHandler.js), right?
```
if (prop === 'smartMode') {
...
		SensiboApi.setDeviceClimateReactState(...)
...
	return
}
```

Also, assuming I got the above correct, does the `return` at the end of the `(prop === 'smartMode')` section need to be removed to ensure both `sensiboApi.setDeviceClimateReactState` line 82 and `sensiboApi.setDeviceACState` line 138 get run (or does the set in StateHandler.js get called twice?)

seidnerj replied:
No worries. Yes, you understood correctly.

There's no need to remove the return (it was there before I got here btw), since when activating Climate React does not necessitate updating the AC state. If necessary, Climate React will update the AC state on its own.

Hope that clears things up.

benwebbbenwebb commented:
Hello @seidnerj 

I've done the merge.

I do have a concern though... I know the current SmartMode (ClimateReact) state will get pulled from Sensibo API every refresh (90ish seconds).

However, what happens if during the 90 second window the user changes their AC state via app? I believe the **_local_** copy of state.smartMode (e.g. acState in unified.js) will then get overwritten by `updateClimateReact()` in StateManager.js... called from within the set functions in StateManager.js (e.g. ACActive). That would update the **local** state.smartMode to new values, right?

And then doesn't that state.smartMode change get **_pushed_** back to Sensibo via `prop === 'smartMode'` in StateHandler.js?

I know it won't turn SmartMode (ClimateReact) **_on or off_** without explicit user action via `ClimateReactEnabledSwitch`, but I think it will still change the SmartMode settings which could be unexpected from the user point of view.

I've also noted that after this merge in the master (main) branch `enableClimateReactAutoSetup` doesn't appear to be used anywhere...

seidnerj commented:
You're right about enableClimateReactAutoSetup not being used, it was probably lost in one of the merges. I opened a new PR with the appropriate fix. Regarding the other issue you mentioned - I will take a look and update this thread (as well as the new PR I already opened).

seidnerj commented:
I thought about what you wrote re the refresh, I'm not sure I completely  followed but assuming I did, I don't see a problem here. Assuming climate react auto setup is enabled, there are two scenarios here:

1. AC state is updated via the plugin --> updateClimateReact() --> state.smartMode gets pushed back to Sensibo --> state.SmartMode's local copy and remote copy are in sync.

2. AC state is updated via the app --> (after ~90s max) refresh happens --> state.SmartMode's local copy and remote copy are in sync.

A state changed that's done via the app will not affect the plugin's local copy of the state.SmartMode until ~90 seconds (max) have passed. When that eventually happens, updateClimateReact() will not be executed because it is not triggered by refreshState() since refreshState() does not update the state via the setters in StateManager.js but rather simply overwrites the local copy with a new object. So, in essence, a state changed triggered via the app does not trigger updateClimateReact() and so state.SmartMode local's copy will not be overwritten.

I think if anything, updateClimateReact() should in fact probably get called after every refresh so that any Climate React's settings would get updated to the new AC state that got pulled from Sensibo's servers in case a user opted to make changes to the AC state via the app vs. the plugin.

Does the above make sense?

benwebbbenwebb commented:
Hi @seidnerj 

I think you and I are on the same page but I was talking about the reverse situation (local state change being sent to Sensibo ClimateReact unexpectedly), which you should now have resolved via #123 

Thanks!

I have a few package version bumps etc in dev-v2.5 branch that I’ll finish soon and then make an alpha release.

benwebbbenwebb commented:
Hi @seidnerj 

I've made the alpha release with your ClimateReact changes, please take a look when you have some time.

seidnerj commented:
Sure, anything specific you need want me to look at?

benwebbbenwebb commented:
> Sure, anything specific you need want me to look at?

Mostly that your new features are working as expected before I promote to beta / release.

Thanks @seidnerj

seidnerj commented:
@benwebbbenwebb I'm running v2.5.0.alpha.2, so far no issues. I will keep you posted.

seidnerj commented:
I haven't noticed any issue so far though I don't use the AC much during the winter time. Have you been using it? Noticed anything "off"?
